### PR TITLE
SALTO-2588 zendesk article attachments

### PR DESCRIPTION
**Pull request by Eldardar**
Add the ability to fetch & deploy article_attachments.
Article attachments are per article and can’t be shared across articles.

To Do:
- [x] UT
- [x] Support in multi-env article body images
- [x] Update creation flow to unassociated article attachments

---

_Additional context for reviewer_

---
_Release Notes_: 
zendesk-adapter:
* Support in fetch & deploy (addition/removal) of article attachments

---
_User Notifications_: 
_Replace me with a short sentence that describes changes that will appear in NaCls and are not caused by user actions (e.g. a new annotation, field values that are converted to references, etc). Hidden changes should not be listed._

### CODE REVIEW

File: packages/zendesk-adapter/src/adapter.ts [hunk 1]
```diff
@@ -416,7 +422,17 @@ export default class ZendeskAdapter implements AdapterOperations {
 
     log.debug('going to run filters on %d fetched elements', elements.length)
     progressReporter.reportProgress({ message: 'Running filters for additional information' })
-    const result = await (await this.createFiltersRunner()).onFetch(elements) as FilterResult
+    const subdomainsWithHelpCenter = elements
+      .filter(isInstanceElement)
+      .filter(instance => instance.elemID.typeName === BRAND_TYPE_NAME)
+      .filter(brandInstance => brandInstance.value.has_help_center)
+      .map(brandInstance => brandInstance.value.subdomain)
+    const subdomainToClient = Object.fromEntries(subdomainsWithHelpCenter.map(
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:430]: shouldn't this be by `brand_id`? otherwise all filters will still need to calculate the mapping themselves

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  }
+  return attachmentInstances
+}
+
+const addArticleAttachment = async (
+  client: ZendeskClient,
+  attachmentInstance: InstanceElement,
+): ReturnType<typeof client.post> => {
+  try {
+    const articleId = attachmentInstance.annotations[CORE_ANNOTATIONS.PARENT][0].id
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:175]: please use `getParents` from adapter-utils (or `getParents` and check the length if we can't assume there is a single parent)

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    const attachmentType = createAttachmentType()
+    const articleAttachments = (await Promise.all(articleInstances
+      .map(async article => getArticleAttachments({
+        client: brandIdToClient[article.value.brand],
+        attachmentType,
+        article,
+      })))).flat()
+
+    // Verify article_attachment type added only once
+    const existedAttachmentType = elements
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:209]: english nit: `existed` -> `existing`

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+      elements.push(attachmentType)
+    }
+    elements.push(...articleAttachments)
+  }, 'Article attachment filter'),
+
+  deploy: async (changes: Change<InstanceElement>[]) => {
+    const [relevantChanges, leftoverChanges] = _.partition(
+      changes,
+      change => getChangeData(change).elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME
+        && isAdditionChange(change),
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:222]: 1. what about modifications? 
2. if we don't support removals, please add a comment explaining why :) should we also have a change validator around these?

File: packages/zendesk-adapter/src/filters/article_body.ts [hunk 5]
```diff
@@ -59,19 +62,59 @@ const referenceArticleUrl = ({
...
+}: {
+  imageUrl: string
+  brandInstances: InstanceElement[]
+  articleAttachmentsInstances: InstanceElement[]
+}): TemplatePart[] => {
+  const urlSubdomain = imageUrl.match(BASE_URL_REGEX)?.pop()
+  const urlBrand = brandInstances
+    .find(brandInstance => brandInstance.value.brand_url === urlSubdomain)
+  const urlArticleAttachmentId = imageUrl.match(ARTICLE_ATTACHMENT_ID_URL_REGEX)?.pop()
+  const referencedArticleAttachment = articleAttachmentsInstances
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_body.ts:78]: there can be a lot of these, so we shouldn't scan every time - let's index these once and pass the index to the function?

File: packages/zendesk-adapter/src/filters/article_body.ts [hunk 6]
```diff
@@ -98,11 +141,18 @@ const filterCreator: FilterCreator = () => {
...
       const articleInstances = instances
         .filter(e => e.elemID.typeName === ARTICLE_TYPE_NAME)
+      const articleAttachmentsInstances = instances
+        .filter(e => e.elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME)
       instances
         .filter(instance => ARTICLE_TYPES.includes(instance.elemID.typeName))
         .filter(articleInstance => !_.isEmpty(articleInstance.value[BODY_FIELD]))
         .forEach(articleInstance => (
-          updateArticleBody(articleInstance, brandInstances, articleInstances)))
+          updateArticleBody(
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_body.ts:150]: there now a few parameters with identical types, let's turn these into named parameters?

**Review by netama — COMMENTED** [review]
looks good :) added some comments and questions

File: packages/zendesk-adapter/src/config.ts [hunk 7]
```diff
@@ -2096,6 +2116,7 @@ export const GUIDE_TYPES_TO_HANDLE_BY_BRAND = [
   'article_translation',
   'category_translation',
   'section_translation',
+  'article_attachment',
```
[the remark this answers is not in the document]
Comment (Eldardar) [reply packages/zendesk-adapter/src/config.ts:2119]: @netama We previously discussed about it (on the `brand` type). I experienced confusion between const and strings when I searched in the config.ts file, so we might want to leave all the type names as strings (or change them all to consts). WDYT?

File: packages/zendesk-adapter/src/filters/article_attachments.ts (hunk 2 above)
Comment (Eldardar) [reply packages/zendesk-adapter/src/filters/article_attachments.ts:175]: @netama This will cause a problem because the parent is already resolved. We need to run this filter after resolve process because we want to assign a brand to each instance.
We can discuss f2f about it?

File: packages/zendesk-adapter/src/config.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/zendesk-adapter/src/config.ts:2165]: eventually IMO we should switch to constants in the config as well, but for now let's just use the new constants we're defining - everything in the code that applies to the same type should use a shared constant so that we can both avoid typos, and rename it more easily if we decide to in the future

File: packages/zendesk-adapter/src/filters/article_attachments.ts (hunk 2 above)
Comment (netama) [reply packages/zendesk-adapter/src/filters/article_attachments.ts:175]: sure. but either way the bigger concern here is that we shouldn't access item `0` before confirming that it exists, as this can cause a crash - so we can use `getParents()` and check the length

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  }
+  return createAttachmentInstance({ attachment, attachmentType, article, content })
+}
+
+const getArticleAttachments = async ({ client, article, attachmentType }: {
+  client: ZendeskClient
+  article: InstanceElement
+  attachmentType: ObjectType
+}): Promise<InstanceElement[]> => {
+  // !!! - pagination
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:143]: leftover?

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+  article: InstanceElement
+  attachmentType: ObjectType
+}): Promise<InstanceElement | undefined> => {
+  const res = await client.getSinglePage({
+    url: `/hc/article_attachments/${attachment.id}/${attachment.file_name}`,
+    responseType: 'arraybuffer',
+  })
+  const content = _.isString(res.data) ? Buffer.from(res.data) : res.data
+  if (!Buffer.isBuffer(content)) {
+    log.error(`Received invalid response from Zendesk API for attachment content, ${safeJsonStringify(res.data, undefined, 2)}. Not adding article attachments`)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:132]: same as the discussion on brand logos, I don't think we should print the attachment buffer here, it can be huge. I think we can omit it completely, if you want to add something please convert to base64 and truncate like you did [there](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/filters/brand_logo.ts#L90).

Ideally we should also reuse the code, but since there are a few more steps here we can keep it this way for now

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  // !!! - pagination
+  const listAttachmentsResponse = await client.getSinglePage({
+    url: `/api/v2/help_center/articles/${article.value.id}/attachments`,
+  })
+  if (listAttachmentsResponse === undefined) {
+    log.error('Received an empty response from Zendesk API. Not adding article attachments')
+    return []
+  }
+  if (Array.isArray(listAttachmentsResponse.data)) {
+    log.error(`Received aninvalid response from Zendesk API, ${safeJsonStringify(listAttachmentsResponse.data, undefined, 2).slice(0, RESULT_MAXIMUM_OUTPUT_SIZE)}. Not adding article attachments`)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:152]:
```suggestion
    log.error(`Received an invalid response from Zendesk API, ${safeJsonStringify(listAttachmentsResponse.data, undefined, 2).slice(0, RESULT_MAXIMUM_OUTPUT_SIZE)}. Not adding article attachments`)
```

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    article.value[ATTACHMENTS_FIELD_NAME] = attachmentInstances
+      .map(instance => new ReferenceExpression(instance.elemID, instance))
+  }
+  return attachmentInstances
+}
+
+const addArticleAttachment = async (
+  client: ZendeskClient,
+  attachmentInstance: InstanceElement,
+): ReturnType<typeof client.post> => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:173]: let's define the type explicitly here, `typeof client.post` is an internal detail of the client, if nothing is relying on it we can return `Promise<void>` and if something is let's define what part it needs

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+  try {
+    const attachmentParents = getParents(attachmentInstance)
+    if (attachmentParents.length !== 1) {
+      throw new Error(`Expected ${attachmentInstance.elemID.getFullName()} to have exactly one parent, found ${attachmentParents.length}`)
+    }
+    const articleId = attachmentParents[0].id
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:179]: missed this before, so just making sure - does this work when creating a new article? we should have something that adds the internal id of the article to the article instance before resolving these changes.
let's also add a check that this is not undefined + a test, to be on the safe side - if we allow here undefined ids we might end up adding the wrong attachment to the wrong article

File: packages/zendesk-adapter/src/filters/article_body.ts [hunk 13]
```diff
@@ -100,11 +150,20 @@ const filterCreator: FilterCreator = () => {
         .filter(e => e.elemID.typeName === BRAND_TYPE_NAME)
       const articleInstances = instances
         .filter(e => e.elemID.typeName === ARTICLE_TYPE_NAME)
+      const articleAttachmentsById: Record<string, InstanceElement> = _.keyBy(
+        instances.filter(e => e.elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME),
+        instance => _.toString(instance.value.id)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_body.ts:155]: same as above - let's make sure we don't index here attachments with undefined ids (+ in parallel, populate the attachment ids from the replies). maybe this is already working :) just couldn't see where we do it yet

**Review by netama — COMMENTED** [review]
looks good, added some nits + some questions about adding articles+attachments at the same deploy

File: packages/zendesk-adapter/src/config.ts [hunk 14]
```diff
@@ -1688,6 +1692,22 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
       },
     },
   },
+  [ARTICLE_ATTACHMENT_TYPE_NAME]: {
+    transformation: {
+      idFields: ['&brand', 'filename'],
```
Comment (netama) [inline packages/zendesk-adapter/src/config.ts:1697]: missed this earlier - we should use the article id here, otherwise we'll have conflicts on filenames (discussing on slack as well, just writing here so we don't miss it)

File: packages/zendesk-adapter/src/adapter.ts [hunk 15]
```diff
@@ -168,6 +169,7 @@ export const DEFAULT_FILTERS = [
   // everyoneUserSegementFilter needs to be before articleFilter
   everyoneUserSegementFilter,
   articleFilter,
+  // articleAttachmentsFilter,
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:172]: can remove?

File: packages/zendesk-adapter/src/config.ts [hunk 16]
```diff
@@ -1649,13 +1649,13 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
     transformation: {
       idFields: ['&section_id', 'title'],
       fileNameFields: ['&section_id', 'title'],
-      standaloneFields: [{ fieldName: 'translations' }],
+      standaloneFields: [{ fieldName: 'translations' }, { fieldName: 'attachments' }],
```
Comment (netama) [inline packages/zendesk-adapter/src/config.ts:1652]: nit:
```suggestion
      standaloneFields: [
        { fieldName: 'translations' },
        { fieldName: 'attachments' },
      ],
```

File: packages/zendesk-adapter/src/config.ts [hunk 17]
```diff
@@ -1649,13 +1649,13 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
...
       sourceTypeName: 'articles__articles',
       fieldsToHide: FIELDS_TO_HIDE.concat(
         { fieldName: 'id', fieldType: 'number' },
         { fieldName: 'position', fieldType: 'number' },
       ),
-      fieldTypeOverrides: [{ fieldName: 'id', fieldType: 'number' }, { fieldName: 'author_id', fieldType: 'unknown' }],
+      fieldTypeOverrides: [{ fieldName: 'id', fieldType: 'number' }, { fieldName: 'author_id', fieldType: 'unknown' }, { fieldName: 'attachments', fieldType: 'list<unknown>' }],
```
Comment (netama) [inline packages/zendesk-adapter/src/config.ts:1658]: nit:
```suggestion
      fieldTypeOverrides: [
        { fieldName: 'id', fieldType: 'number' },
        { fieldName: 'author_id', fieldType: 'unknown' },
        { fieldName: 'attachments', fieldType: 'list<unknown>' },
      ],
```
(also can we make these not unknown, e.g. make `author_id` a number/string?

File: packages/zendesk-adapter/src/config.ts (hunk 14 above)
Comment (netama) [reply packages/zendesk-adapter/src/config.ts:1696]: reopening as it doesn't look like this was addressed - we should make this a child of the parent, so probably
```suggestion
      idFields: ['filename'],
      extendsParentId: true,
```
and make sure the article is its parent so that the id gets re-generated correctly in the `reference_id_fields` filter

File: packages/zendesk-adapter/src/filters/article.ts [hunk 18]
```diff
@@ -67,87 +71,173 @@ const setUserSegmentIdForAdditionChanges = (
     })
 }
 
+const haveAttachmentsBeenAdded = (
+  articleChange: AdditionChange<InstanceElement> | ModificationChange<InstanceElement>
+): boolean => {
+  const addedAttachments = isAdditionChange(articleChange)
+    ? articleChange.data.after.value.attachments
+    : _.difference(
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:79]: we shouldn't compare objects "by pointer", can we use `_.differenceBy()` or `_.differenceWith()` to make sure we don't mis-calculate this?
also, do we need to do something if the attachments are identical but ordered differently?

File: packages/zendesk-adapter/src/filters/article.ts [hunk 19]
```diff
@@ -67,87 +71,173 @@ const setUserSegmentIdForAdditionChanges = (
...
+      articleChange.data.after.value.attachments,
+      articleChange.data.before.value.attachments,
+    )
+  if (!_.isArray(addedAttachments)) {
+    return false
+  }
+  return addedAttachments.length > 0
+}
+
+const AssociateAttachments = async (
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:89]:
```suggestion
const associateAttachments = async (
```

File: packages/zendesk-adapter/src/filters/article.ts [hunk 20]
```diff
@@ -67,87 +71,173 @@ const setUserSegmentIdForAdditionChanges = (
...
+          article.value[USER_SEGMENT_ID_FIELD] = new ReferenceExpression(
+            everyoneUserSegmentInstance.elemID,
+            everyoneUserSegmentInstance,
           )
-        } catch (e) {
-          log.error('Error parsing article body value in deployment', e)
-        }
-      })
-  },
+        })
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:130]: nit: maybe move this to a helper function so that the `onFetch` logic is easier to follow?

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+      }),
+      inline: attachment.inline,
+      brand: article.value.brand,
+    },
+    [ZENDESK, RECORDS_PATH, ARTICLE_ATTACHMENT_TYPE_NAME, pathName],
+    { [CORE_ANNOTATIONS.PARENT]: [new ReferenceExpression(article.elemID, article)] },
+  )
+}
+
+export const createAttachmentType = (): ObjectType =>
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:102]: if we're creating it here do we really need the configuration as well? is anything using it?

File: packages/zendesk-adapter/src/filters/article.ts [hunk 22]
```diff
@@ -67,87 +71,173 @@ const setUserSegmentIdForAdditionChanges = (
...
-        removeTitleAndBody(articleInstance)
-        if (articleInstance.value[USER_SEGMENT_ID_FIELD] === null) {
-          articleInstance.value[USER_SEGMENT_ID_FIELD] = new ReferenceExpression(
-            everyoneUserSegmentInstance.elemID,
-            everyoneUserSegmentInstance,
+      // Verify article_attachment type added only once
+      const existingAttachmentType = elements
+        .filter(isObjectType)
+        .find(objType => objType.elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME)
+      if (existingAttachmentType === undefined) {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:144]: if the type already exists shouldn't we have used it when creating the article instances above?

File: packages/zendesk-adapter/src/filters/article.ts [hunk 23]
```diff
@@ -67,87 +71,173 @@ const setUserSegmentIdForAdditionChanges = (
...
+        .map(getChangeData)
+      await awu(addedArticleAttachments)
+        .forEach(async attachmentInstance => {
+          await createUnassociatedAttachment(client, attachmentInstance)
+          // Keeping article-attachment relation for deploy stage
+          const unresolvedInstance = await elementsSource.get(attachmentInstance.elemID)
+          if (unresolvedInstance === undefined) {
+            return
+          }
+          const parentArticleRef = unresolvedInstance.annotations[CORE_ANNOTATIONS.PARENT][0]
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:164]: please use `getParent` (and handle the potential exception)

File: packages/zendesk-adapter/src/filters/article_attachments.ts [hunk 24]
```diff
@@ -0,0 +1,287 @@
+/*
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article_attachments.ts:1]: not sure what parts of this are still relevant and what's getting removed?

File: packages/zendesk-adapter/src/filters/brand_logo.ts [hunk 25]
```diff
@@ -81,7 +81,7 @@ const getLogoContent = async (
   logoFileName: string,
 ): Promise<Buffer | undefined> => {
   const res = await client.getResource({
-    url: `/brands/${logoId}/${logoFileName}`,
+    url: `/system/brands/${logoId}/${logoFileName}`,
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brand_logo.ts:84]: this is not related, right? can you please add a separate PR for it so that it's fixed ASAP, since it's currently broken?

File: packages/zendesk-adapter/src/filters/article.ts (hunk 23 above)
Comment (Eldardar) [reply packages/zendesk-adapter/src/filters/article.ts]: Discussed f2f

File: packages/zendesk-adapter/src/filters/article_attachments.ts (hunk 21 above)
Comment (Eldardar) [reply packages/zendesk-adapter/src/filters/article_attachments.ts:102]: @netama re - we need the configuration for deploying 'removal' changes

File: packages/zendesk-adapter/src/filters/article.ts [hunk 26]
```diff
@@ -67,87 +92,164 @@ const setUserSegmentIdForAdditionChanges = (
...
+      // Creating unassociated article attachments
+      const addedArticleAttachments = changes
+        .filter(isAdditionChange)
+        .filter(change => getChangeData(change).elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME)
+        .map(getChangeData)
+      await awu(addedArticleAttachments)
+        .forEach(async attachmentInstance => {
+          await createUnassociatedAttachment(client, attachmentInstance)
+          // Keeping article-attachment relation for deploy stage
+          const unresolvedInstance = await elementsSource.get(attachmentInstance.elemID)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:165]: nit: `unresolved` here refers to the references not being resolved, right? this is a little confusing because usually it means we couldn't find the instance itself (like "unresolved references"), maybe `instanceBeforeResolve`?

File: packages/zendesk-adapter/src/filters/article.ts [hunk 27]
```diff
@@ -67,87 +92,164 @@ const setUserSegmentIdForAdditionChanges = (
...
+          if (unresolvedInstance === undefined) {
+            return
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:167]: is this an expected scenario? if not let's log?

File: packages/zendesk-adapter/src/filters/article.ts [hunk 28]
```diff
@@ -67,87 +92,164 @@ const setUserSegmentIdForAdditionChanges = (
...
+          }
+          const parentArticleList = unresolvedInstance.annotations[CORE_ANNOTATIONS.PARENT]
+          if (!_.isArray(parentArticleList)) {
+            return
+          }
+          const parentArticleRef = parentArticleList[0]
+          if (!isReferenceExpression(parentArticleRef)) {
+            return
+          }
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/article.ts:176]: nit: let's move this to a helper function?

**Review by netama — APPROVED** [review]
very cool :) added some comments, pre-approving to unblock for after the unittests are added
